File: app/src/main/java/com/simsimhan/promissu/promise/PromiseFragment.java
package com.simsimhan.promissu.promise;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.Toast;

import com.simsimhan.promissu.MainActivity;
import com.simsimhan.promissu.PromissuApplication;
import com.simsimhan.promissu.R;
import com.simsimhan.promissu.network.AuthAPI;
import com.simsimhan.promissu.network.model.Promise;
import com.simsimhan.promissu.util.NavigationUtil;

import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class PromiseFragment extends Fragment implements SwipeRefreshLayout.OnRefreshListener {
    public static final String TAG = "PromiseFragment";
    private CompositeDisposable disposables;
    private String token;
    private SwipeRefreshLayout swipeContainer;
    private PromiseAdapter adapter;
    private RecyclerView recyclerView;
    private Toolbar toolbar;
    private ImageView emptyImageView;
    private boolean isPastPromise;
    private ConstraintLayout emptyHolder;

    public static PromiseFragment newInstance(int position, String title, boolean isPastPromise) {
        PromiseFragment fragment = new PromiseFragment();
        Bundle args = new Bundle();
        args.putInt("Page_key", position);
        args.putString("Title_key", title);
        args.putBoolean("is_past_key", isPastPromise);
        fragment.setArguments(args);
        return fragment;
    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getArguments() != null) {
            // get arguments and set here
            isPastPromise = getArguments().getBoolean("is_past_key");
        }

        token = PromissuApplication.Companion.getDiskCache().getUserToken();
        adapter = new PromiseAdapter((AppCompatActivity) getActivity(), new ArrayList<>(), isPastPromise);
        disposables = new CompositeDisposable();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_promise_list, container, false);

        recyclerView = view.findViewById(R.id.promise_recycler_view);
        swipeContainer = view.findViewById(R.id.swipe_container);
        emptyHolder = view.findViewById(R.id.empty_view_holder);
        emptyImageView = view.findViewById(R.id.empty_view_image);

        emptyHolder.setVisibility(View.GONE);
        swipeContainer.setOnRefreshListener(this);
        swipeContainer.setColorSchemeResources(R.color.colorPrimary, android.R.color.holo_green_dark, android.R.color.holo_orange_dark, android.R.color.holo_blue_dark);
        swipeContainer.setProgressViewOffset(false, 0, getResources().getDimensionPixelSize(R.dimen.refresher_start));
        swipeContainer.post(() -> {
            swipeContainer.setRefreshing(true);
            fetch();
        });

        recyclerView.setAdapter(adapter);
        recyclerView.addItemDecoration(new DividerItemDecoration(recyclerView.getContext(), DividerItemDecoration.VERTICAL));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            recyclerView.setOnScrollChangeListener((v, scrollX, scrollY, oldScrollX, oldScrollY) -> {
                if (toolbar != null && recyclerView != null) {
                    toolbar.setSelected(recyclerView.canScrollVertically(-1));
                }
            });
        }

        return view;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        Timber.d("onActivityResult(): " + requestCode + " " + resultCode);

    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        // inflate menuinflator if we need action from options
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        disposables.clear();
    }

    private void fetch() {
        disposables.add(
                PromissuApplication.Companion.getRetrofit()
                        .create(AuthAPI.class)
                        .getMyPromise("Bearer " + token, 0, 9, isPastPromise ? "past" : "future")
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(onNext -> {
                            swipeContainer.setRefreshing(false);
                            adapter.reset(onNext);
                            setEmptyViewVisible(onNext.size() <= 0);
                        }, onError -> {
                            swipeContainer.setRefreshing(false);
                            Timber.e(onError);
                            Toast.makeText(getActivity(), "서버 점검 중입니다. 다시 로그인해 주세요.", Toast.LENGTH_SHORT).show();
                            PromissuApplication.Companion.getDiskCache().clearUserData();
                            NavigationUtil.replaceWithLoginView((AppCompatActivity) getActivity());
                        }));
    }

    private void setEmptyViewVisible(boolean setVisible) {
        if (emptyHolder != null && recyclerView != null && emptyImageView != null) {
//            if (setVisible) recyclerView.scrollTo(0, 0);
            emptyHolder.setVisibility(setVisible ? View.VISIBLE : View.INVISIBLE);
            recyclerView.setBackgroundColor(ContextCompat.getColor(recyclerView.getContext(), isPastPromise ? R.color.past_background_color : R.color.background_grey));
            emptyImageView.setImageDrawable(ContextCompat.getDrawable(emptyImageView.getContext(), isPastPromise ? R.drawable.no_appointment_red : R.drawable.no_appointment_blue));
            emptyHolder.setBackgroundColor(ContextCompat.getColor(emptyHolder.getContext(), isPastPromise ? R.color.past_background_color : R.color.background_grey));
        }
    }

    @Override
    public void onRefresh() {
        fetch();
    }
}

File: app/src/main/java/com/simsimhan/promissu/promise/PromiseFragment.kt
package com.simsimhan.promissu.promise

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.simsimhan.promissu.BuildConfig
import com.simsimhan.promissu.PromissuApplication
import com.simsimhan.promissu.R
import com.simsimhan.promissu.databinding.FragmentPromiseListBinding
import com.simsimhan.promissu.network.AuthAPI
import com.simsimhan.promissu.util.NavigationUtil
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import org.koin.androidx.viewmodel.ext.android.sharedViewModel

class PromiseFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {


    private lateinit var binding: FragmentPromiseListBinding
    private lateinit var adapter: PromiseAdapter
    private val viewModel: PromiseViewModel by sharedViewModel()
    private var isPastPromise = false
    private val disposables = CompositeDisposable()

    companion object {
        fun newInstance(position: Int, title: String, isPastPromise: Boolean): PromiseFragment {
            val fragment = PromiseFragment()
            val args = Bundle()
            args.putInt("Page_key", position)
            args.putString("Title_key", title)
            args.putBoolean("is_past_key", isPastPromise)
            fragment.arguments = args
            return fragment
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (arguments != null) {
            isPastPromise = arguments!!.getBoolean("is_past_key")
        }

//        adapter = PromiseAdapter(activity as AppCompatActivity?, arrayListOf(), isPastPromise)
        adapter = PromiseAdapter(arrayListOf(), isPastPromise, this@PromiseFragment, this.viewModel)

        viewModel.errToastMsg.observe(this, Observer {
            binding.swipeContainer.isRefreshing = false
            toastMessage(it)
            NavigationUtil.replaceWithLoginView(activity as AppCompatActivity)
        })

        if(!isPastPromise) {
            viewModel.deleteRoom.observe(this, Observer {
                buildDeleteDialog(it.id)
            })
        }

        viewModel.toastMsg.observe(this, Observer {
            toastMessage(it)
        })


    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        binding = FragmentPromiseListBinding.inflate(inflater, container, false).apply {
            lifecycleOwner = this@PromiseFragment
            viewModel = this@PromiseFragment.viewModel
            promiseRecyclerView.apply {
                adapter = this@PromiseFragment.adapter
                addItemDecoration(DividerItemDecoration(this.context, DividerItemDecoration.VERTICAL))
            }
            swipeContainer.apply {
                setOnRefreshListener(this@PromiseFragment)
                setColorSchemeResources(R.color.colorPrimary, android.R.color.holo_green_dark, android.R.color.holo_orange_dark, android.R.color.holo_blue_dark)
                setProgressViewOffset(false, 0, resources.getDimensionPixelSize(R.dimen.refresher_start))
                post {
                    swipeContainer.isRefreshing = true
                    fetch(isPastPromise)
                }
            }
        }

        if (isPastPromise) {
            viewModel.onListLoadedPast.observe(this, Observer {
                binding.swipeContainer.isRefreshing = false
                adapter.reset(it)
                setEmptyViewVisible(it.isEmpty())
            })
        } else {
            viewModel.onListLoadedRecent.observe(this, Observer {
                binding.swipeContainer.isRefreshing = false
                adapter.reset(it)
                setEmptyViewVisible(it.isEmpty())
            })
        }

        return binding.root
    }

    private fun setEmptyViewVisible(setVisible: Boolean) {
        binding.apply {
            emptyViewHolder.visibility = if (setVisible) View.VISIBLE else View.INVISIBLE
            promiseRecyclerView.setBackgroundColor(ContextCompat.getColor(promiseRecyclerView.context, if (isPastPromise) R.color.past_background_color else R.color.background_grey))
            emptyViewImage.setImageDrawable(ContextCompat.getDrawable(emptyViewImage.context, if (isPastPromise) R.drawable.no_appointment_red else R.drawable.no_appointment_blue))
            emptyViewHolder.setBackgroundColor(ContextCompat.getColor(emptyViewHolder.context, if (isPastPromise) R.color.past_background_color else R.color.background_grey))
            emptyViewText.setTextColor(ContextCompat.getColor(emptyViewText.context, if (isPastPromise) R.color.past_background_dark else R.color.colorStrong))
        }
    }

    override fun onResume() {
        super.onResume()
        fetch(isPastPromise)
    }

    override fun onPause() {
        super.onPause()
        disposables.clear()
    }

    override fun onRefresh() {
        fetch(isPastPromise)
    }

    private fun fetch(isPastPromise: Boolean) {
        viewModel.fetch(isPastPromise)
    }

    private fun toastMessage(msg: String) {
        Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
    }

    private fun buildDeleteDialog(room_id: Int) {
        val dialog = Dialog(context)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(R.layout.dialog_request)
        val text1 = dialog.findViewById<TextView>(R.id.dialog_text1)
        val text2 = dialog.findViewById<TextView>(R.id.dialog_text2)
        val btnCancel = dialog.findViewById<Button>(R.id.dialog_button_cancel)
        val btnAccept = dialog.findViewById<Button>(R.id.dialog_button_accept)
        text1.text = "정말 약속을"
        text2.text = "나가시겠습니까?"
        btnAccept.text = "나가기"
        btnCancel.text = "취소"
        btnAccept.setOnClickListener {
            // successfully delete room
            disposables.add(
                    PromissuApplication.retrofit!!
                            .create(AuthAPI::class.java)
                            .leftAppointment("Bearer " + PromissuApplication.diskCache!!.userToken, room_id)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe({ onNext ->
                                if (onNext.code() == 200) {
                                    //
                                    if (!BuildConfig.DEBUG) {
                                        val eventParams = Bundle()
                                        eventParams.putInt("room_id", room_id)
                                        eventParams.putLong("user_id", PromissuApplication.diskCache!!.userId)
                                        PromissuApplication.firebaseAnalytics!!.logEvent("appointment_left", eventParams)
                                    }
                                    fetch(isPastPromise)
                                    dialog.dismiss()
                                } else if (onNext.code() == 401) {
                                    Toast.makeText(context, "삭제 권한이 없습니다.", Toast.LENGTH_SHORT).show()
                                } else {
                                    Toast.makeText(context, "삭제에 실패 했습니다. 잠시후 다시 시도해주세요.", Toast.LENGTH_SHORT).show()
                                }
                            }, {
                                if (BuildConfig.DEBUG) {
                                    Toast.makeText(context, "삭제 에러", Toast.LENGTH_SHORT).show()
                                }
                                dialog.dismiss()
                            }))
        }
        btnCancel.setOnClickListener { dialog.dismiss() }
        dialog.setCancelable(false)
        dialog.show()
    }

}